### PR TITLE
feat: モデルの定義

### PR DESCRIPTION
**Pull request by maton369**

### CODE REVIEW

**Review by maton369 — COMMENTED** [review]
## レビューコメント（PropensityMatching の設定：理論的な観点から）

### 1. 全体像：いま作っている推定器は何をしているか

- 表示されているオブジェクトは

  PropensityMatching(
      learner=LogisticRegression(class_weight='balanced', solver='liblinear'),
      propensity_transform=PropensityTransformer(learner=LogisticRegression(...)),
      metric='mahalanobis',
      n_neighbors=1,
      matching_mode='both',
      with_replacement=True,
      caliper=None,
      covariance_conditioner=EmpiricalCovariance(),
      estimate_observed_outcome=False,
      knn_backend='sklearn',
  )

  という構成になっている。
- 理論的にはこれは、

  1. ロジスティック回帰で傾向スコア  
     e(X) = P(A=1 | X)

     を推定し、
  2. 傾向スコア空間＋共変量空間における **距離（マハラノビス距離）** を使って
     A=1 と A=0 を 1 対 1（n_neighbors=1）でマッチングし、
  3. マッチされたペアのアウトカム差から因果効果（主に ATT）を推定する

  という「典型的な傾向スコアマッチング推定器」になっている。

- すでに `learner=LogisticRegression(class_weight='balanced', solver='liblinear')`
  という設定で、クラス不均衡（禁煙者/非禁煙者の比率）に配慮した
  傾向スコア推定モデルを構築しているので、
  **大きな理論的破綻はなく、教科書的な設定**と言える。

---

### 2. 傾向スコアマッチングの理論的位置づけ

#### 2-1. ポテンシャルアウトカム・フレームワーク

- NHEFS のような観察データでは、各個体 i について

  - Aᵢ ∈ {0,1}: 処置（禁煙したかどうか）
  - Yᵢ: 観測アウトカム（体重変化）
  - Xᵢ: 交絡因子を含む共変量

  が観測される。
- 真に興味があるのは

  - Yᵢ(1): 「i が禁煙した世界」での体重変化
  - Yᵢ(0): 「i が禁煙しなかった世界」での体重変化

  といった「反事実」も含むポテンシャルアウトカムであり、
  真の因果効果（例えば ATE, ATT）は

  - ATE = E[Y(1) - Y(0)]
  - ATT = E[Y(1) - Y(0) | A=1]

  のように定義される。

#### 2-2. 傾向スコア e(X) の役割

- 観察データしかない状況で因果効果を識別するには、
  少なくとも「交絡因子で条件付けしたときの交換可能性」

  - {Y(0), Y(1)} ⟂ A | X

  を仮定し、その上で X を「うまく調整する」必要がある。
- Rosenbaum & Rubin の理論によれば、傾向スコア

  - e(X) = P(A=1 | X)

  は条件付き独立性を保つ十分統計量であり、

  - {Y(0), Y(1)} ⟂ A | e(X)

  が成り立つ（交換可能性が X ではなく e(X) で言い換えられる）ので、
  **フルの共変量ベクトル X ではなく、1 次元の e(X) に基づくマッチングや層化でも
  交絡を制御しうる**ことが理論的に保証される。

- PropensityMatching はまさにこの理論に基づき、

  - 「似た e(X) を持つ A=1 と A=0 の個体をペアリングする」

  ことで、「処置割付がほぼランダムな疑似試験」を作ろうとしている。

---

### 3. 個々のハイパーパラメータの理論的な意味

#### 3-1. learner / propensity_transform

- `learner=LogisticRegression(...)`
- `propensity_transform=PropensityTransformer(learner=LogisticRegression(...))`

  という二重指定は、内部的には

  - 「傾向スコア推定を行うロジスティック回帰」
  - 「それを変換器としてパイプラインに組み込む」

  という構造を示している。

- ロジスティック回帰は

  logit(e(X)) = β₀ + βᵀ X

  と仮定するモデルであり、線形予測子に対するロジスティックリンク関数を通じて
  e(X) を推定する。
- 因果推論の観点では、

  - (1) e(X) を十分によく近似していること
  - (2) e(X) と X が高次元でも過学習しすぎないこと（極端な 0/1 を出さない）
  が重要で、ロジスティック回帰はそのバランスから「最初の選択肢」として妥当。

- `class_weight="balanced"` により、
  - 少数クラス（禁煙者 or 非禁煙者）が極端に無視されず、
  - e(X) 推定におけるバイアスをある程度抑えようとしている点も理にかなっている。

#### 3-2. metric='mahalanobis' と covariance_conditioner=EmpiricalCovariance()

- `metric='mahalanobis'` は、マッチングの距離として **マハラノビス距離** を使う指定。

  - d_M(xᵢ, xⱼ) = √{ (xᵢ - xⱼ)ᵀ Σ⁻¹ (xᵢ - xⱼ) }

  であり、共分散 Σ を用いてスケールと相関構造を補正したユークリッド距離に相当する。

- マッチング理論の観点から見ると、

  - 単純なユークリッド距離（スケール依存）よりも、
    各共変量のスケールや相関を考慮したマハラノビス距離の方が、
    「共変量バランス」を取りやすいことが知られている。
  - 特に多次元 X を扱う場合、各変数の単位や分散が異なるため、
    マハラノビス距離で標準化するのは自然な選択。

- `covariance_conditioner=EmpiricalCovariance()` は Σ を推定する手段で、
  単純にサンプル共分散を用いて Σ を推定し、それを距離計算に使う構成。
  共分散行列がフルランクであり、サンプル数が十分に多いときには standard な選択。

#### 3-3. n_neighbors=1, matching_mode='both', with_replacement=True

- `n_neighbors=1`:
  - これは「1 対 1 マッチング（nearest neighbor matching）」を意味する。
  - 理論的には、各処置個体に最も似た対照個体を 1 つだけ割り当てることで、
    シンプルかつ解釈しやすい推定量（ATT）を得ることができる。
  - ただし、バラツキの大きいデータでは「悪いマッチ」（距離が離れたペア）が
    混じることもあるので、後述の caliper などとの併用が重要になる。

- `matching_mode='both'`:
  - 通常、マッチングは
    - 処置群（A=1）の個体を基準にして対照群（A=0）を探す「treated on control」
    - その逆「control on treated」
    のいずれかで行うが、
  - 'both' は両方向を考慮するモードで、
    - 実装上は「どちらが基準でもよい」ように探索し、より良いマッチを作ろうとする設定と理解できる。
  - これにより、特定の群が極端に少ない場合でも、マッチングの柔軟性が増す。

- `with_replacement=True`:
  - 対照群の個体が「複数の処置個体に再利用される」ことを許す設定。
  - 理論的には、
    - 再利用を許すことで「良い対照個体」が複数の処置個体のマッチング相手になりうるため、
      全体として平均的なマッチの質（距離）が向上しやすい。
    - 一方で、特定の対照個体に大きな「重み」が集中するため、分散は増えうる。
  - 再利用を禁止（without replacement）すると、
    - 対照群のカバレッジは広がるが、マッチの質が悪化しやすく、
    - bias-variance tradeoff の観点でどちらが良いかはケースバイケース。
  - 再利用を許した 1 対 1 マッチングは、ATT 推定でよく用いられる構成であり、
    NHEFS のような実データでも一般的に採用されている。

#### 3-4. caliper=None

- `caliper=None` は「キャリパー（距離の許容半径）を設定しない」という意味。
- キャリパー c を設定すると、「距離が c を超えるマッチ」は不適合として捨てることができ、
  - 非常に遠いペア（共変量が全然似ていない処置・対照ペア）を排除し、
  - 残ったサンプルに対してより良い共変量バランスを得ることができる。
- 一方で、キャリパーを厳しくするとサンプルが多く捨てられ、推定の分散が増える（効率が落ちる）。
- `caliper=None` は「まずは距離制約なしで様子を見る」設定として妥当だが、
  - 実務的には、マッチング後の共変量バランス（標準化差分など）をチェックしたうえで、
    必要に応じて caliper を導入・調整するのがよい。

#### 3-5. estimate_observed_outcome=False

- `estimate_observed_outcome=False` は、
  - 「観測アウトカム自体を learner でモデリングして補う」ような処理を
    このクラス内部では行わない、という指定と解釈できる。
- これは、純粋に「マッチングによる再重み付け」によって因果効果を評価したい、
  という立場に整合的であり、
  - g-computation（アウトカムモデルベースの推定）と混ざらない分、
  - 理論的な整理もしやすい。

---

### 4. 改善提案・発展アイデア

1. **キャリパーと共変量バランスの可視化**
   - 現状の設定でも理論的には問題ないが、
     - マッチング前後のバランス（標準化差分、SMD）をプロットし、
     - 必要であれば `caliper` を導入して「悪いマッチ」を削る、
     というループを回すと、より因果推定としての説得力が増す。

2. **ATT / ATE の明示**
   - PropensityMatching のデフォルト設定が主に ATT をターゲットにしているのか、
     ATE も視野に入れているのかをコード・コメントで明示しておくと良い。
   - 一般に 1:1 マッチング with replacement は ATT 志向で用いられるため、
     - 「禁煙者に対する平均処置効果（ATT）を評価している」
     という解釈を README やノートに書いておくと、理論と実装の結びつきが明確になる。

3. **他の推定器との比較**
   - 同じ NHEFS データで
     - IPWEstimator（IPW）
     - Doubly Robust Estimator（DR）
   なども試し、PropensityMatching の結果と比較すると、
     - 「マッチング推定量のバイアスと分散の特徴」
     - 「DR 推定量のロバスト性」
     など、因果推論の理論的ポイントをより立体的に示せる。

---

### 5. まとめ

- いまの PropensityMatching の設定は、
  - ロジスティック回帰による傾向スコア推定
  - マハラノビス距離に基づく 1:1 最近傍マッチング
  - クラス不均衡への配慮（class_weight="balanced"）
  - 対照群の再利用許可（with_replacement=True）
  という構成で、**教科書の理論とよく整合したオーソドックスな選択**になっている。
- ここから先は、
  - マッチング後の共変量バランス評価
  - ATT/ATE の明示
  - IPW/DR との比較
  などを追加することで、リポジトリとして「理論と実装を行き来できる因果推論の教材」としてさらに強くできると思う。
